File: Mk8.Sava.Tests/SdkCompatibility/cpp/main.cpp
#include <azure/storage/blobs.hpp>
#include <azure/storage/blobs/blob_lease_client.hpp>
#include <azure/storage/common/storage_credential.hpp>

#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <memory>
#include <random>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace Blobs = Azure::Storage::Blobs;
namespace Models = Azure::Storage::Blobs::Models;

namespace {

constexpr auto SdkVersion = "12.18.0";

void Require(bool condition, const std::string& message)
{
  if (!condition)
  {
    throw std::runtime_error(message);
  }
}

std::string RequiredEnvironment(const char* name)
{
  auto value = std::getenv(name);
  if (value == nullptr || *value == '\0')
  {
    throw std::runtime_error(std::string("missing required environment variable ") + name);
  }
  return value;
}

std::string RandomHex(std::size_t byteCount)
{
  std::random_device source;
  std::ostringstream output;
  output << std::hex << std::setfill('0');
  for (std::size_t index = 0; index < byteCount; ++index)
  {
    output << std::setw(2) << static_cast<unsigned int>(source() & 0xffU);
  }
  return output.str();
}

std::vector<std::uint8_t> Download(const Blobs::BlobClient& client)
{
  auto response = client.Download();
  return response.Value.BodyStream->ReadToEnd();
}

std::vector<std::uint8_t> DownloadRange(
    const Blobs::BlobClient& client,
    std::int64_t offset,
    std::int64_t length)
{
  Blobs::DownloadBlobOptions options;
  options.Range = Azure::Core::Http::HttpRange{offset, length};
  auto response = client.Download(options);
  return response.Value.BodyStream->ReadToEnd();
}

} // namespace

int main()
{
  const auto endpoint = RequiredEnvironment("MK8_SAVA_BLOB_ENDPOINT");
  const auto accountName = RequiredEnvironment("MK8_SAVA_ACCOUNT_NAME");
  const auto accountKey = RequiredEnvironment("MK8_SAVA_ACCOUNT_KEY");
  auto credential
      = std::make_shared<Azure::Storage::StorageSharedKeyCredential>(accountName, accountKey);
  Blobs::BlobServiceClient service(endpoint, credential);
  const auto containerName = "cpp-" + RandomHex(16);
  auto container = service.GetBlobContainerClient(containerName);
  bool created = false;

  try
  {
    Blobs::CreateBlobContainerOptions createOptions;
    createOptions.Metadata = {{"runtime", "cpp"}};
    container.Create(createOptions);
    created = true;

    Blobs::ListBlobContainersOptions containerListOptions;
    containerListOptions.Prefix = containerName;
    containerListOptions.Include = Models::ListBlobContainersIncludeFlags::Metadata;
    std::size_t containerCount = 0;
    for (auto page = service.ListBlobContainers(containerListOptions); page.HasPage();
         page.MoveToNextPage())
    {
      for (const auto& item : page.BlobContainers)
      {
        Require(item.Name == containerName, "unexpected container name");
        Require(item.Details.Metadata.at("runtime") == "cpp", "container metadata mismatch");
        ++containerCount;
      }
    }
    Require(containerCount == 1, "expected exactly one container");

    std::vector<std::uint8_t> content(256 * 1024 + 53);
    for (std::size_t index = 0; index < content.size(); ++index)
    {
      content[index] = static_cast<std::uint8_t>((index * 43 + index / 277) % 251);
    }

    auto block = container.GetBlockBlobClient("folder/sdk-roundtrip.bin");
    Blobs::UploadBlockBlobFromOptions uploadOptions;
    uploadOptions.HttpHeaders.ContentType = "application/x-mk8-cpp";
    uploadOptions.HttpHeaders.CacheControl = "private,max-age=31";
    uploadOptions.Metadata = {{"owner", "cpp-sdk"}};
    uploadOptions.Tags = {{"runtime", "cpp"}, {"purpose", "compatibility"}};
    block.UploadFrom(content.data(), content.size(), uploadOptions);
    Require(Download(block) == content, "block download mismatch");
    const auto expectedRange = std::vector<std::uint8_t>(
        content.begin() + 12'345,
        content.begin() + 20'122);
    Require(DownloadRange(block, 12'345, 7'777) == expectedRange, "range download mismatch");

    const auto properties = block.GetProperties().Value;
    Require(properties.BlobSize == static_cast<std::int64_t>(content.size()), "content length mismatch");
    Require(properties.HttpHeaders.ContentType == "application/x-mk8-cpp", "content type mismatch");
    Require(properties.HttpHeaders.CacheControl == "private,max-age=31", "cache control mismatch");
    Require(properties.Metadata.at("owner") == "cpp-sdk", "blob metadata mismatch");
    Require(block.GetTags().Value.at("runtime") == "cpp", "blob tag mismatch");

    Blobs::ListBlobsOptions blobListOptions;
    blobListOptions.Prefix = "folder/";
    blobListOptions.Include
        = Models::ListBlobsIncludeFlags::Metadata | Models::ListBlobsIncludeFlags::Tags;
    std::size_t blobCount = 0;
    for (auto page = container.ListBlobs(blobListOptions); page.HasPage(); page.MoveToNextPage())
    {
      for (const auto& item : page.Blobs)
      {
        Require(item.Name == "folder/sdk-roundtrip.bin", "listed blob name mismatch");
        Require(item.Details.Metadata.at("owner") == "cpp-sdk", "listed metadata mismatch");
        Require(item.Details.Tags.at("runtime") == "cpp", "listed tag mismatch");
        ++blobCount;
      }
    }
    Require(blobCount == 1, "expected exactly one listed blob");

    Blobs::CreateBlobSnapshotOptions snapshotOptions;
    snapshotOptions.Metadata = {{"owner", "snapshot"}};
    auto snapshotResult = block.CreateSnapshot(snapshotOptions).Value;
    auto snapshot = block.WithSnapshot(snapshotResult.Snapshot);
    Require(Download(snapshot) == content, "snapshot download mismatch");
    Require(
        snapshot.GetProperties().Value.Metadata.at("owner") == "snapshot",
        "snapshot metadata mismatch");

    auto staged = container.GetBlockBlobClient("staged.bin");
    const std::vector<std::string> blockIds = {
        "YmxvY2stMDAwMQ==",
        "YmxvY2stMDAwMg==",
    };
    const std::vector<std::uint8_t> first = {'f', 'i', 'r', 's', 't', '-', 'b', 'l', 'o', 'c', 'k', '|'};
    const std::vector<std::uint8_t> second = {'s', 'e', 'c', 'o', 'n', 'd', '-', 'b', 'l', 'o', 'c', 'k'};
    Azure::Core::IO::MemoryBodyStream firstStream(first.data(), first.size());
    staged.StageBlock(blockIds[0], firstStream);
    Azure::Core::IO::MemoryBodyStream secondStream(second.data(), second.size());
    staged.StageBlock(blockIds[1], secondStream);
    Blobs::CommitBlockListOptions commitOptions;
    commitOptions.Metadata = {{"route", "staged"}};
    commitOptions.Tags = {{"runtime", "cpp"}};
    staged.CommitBlockList(blockIds, commitOptions);
    auto expectedStaged = first;
    expectedStaged.insert(expectedStaged.end(), second.begin(), second.end());
    Require(Download(staged) == expectedStaged, "staged block download mismatch");
    auto blockList = staged.GetBlockList().Value;
    Require(blockList.CommittedBlocks.size() == blockIds.size(), "committed block count mismatch");
    for (std::size_t index = 0; index < blockIds.size(); ++index)
    {
      Require(blockList.CommittedBlocks[index].Name == blockIds[index], "committed block ID mismatch");
    }
    Require(blockList.UncommittedBlocks.empty(), "unexpected uncommitted blocks");

    auto append = container.GetAppendBlobClient("append.log");
    append.Create();
    const std::vector<std::uint8_t> alpha = {'a', 'l', 'p', 'h', 'a'};
    const std::vector<std::uint8_t> beta = {'|', 'b', 'e', 't', 'a'};
    Azure::Core::IO::MemoryBodyStream alphaStream(alpha.data(), alpha.size());
    append.AppendBlock(alphaStream);
    Azure::Core::IO::MemoryBodyStream betaStream(beta.data(), beta.size());
    append.AppendBlock(betaStream);
    auto expectedAppend = alpha;
    expectedAppend.insert(expectedAppend.end(), beta.begin(), beta.end());
    Require(Download(append) == expectedAppend, "append blob download mismatch");

    auto pageBlob = container.GetPageBlobClient("page.bin");
    pageBlob.Create(1024);
    std::vector<std::uint8_t> pageBytes(512, 0x5a);
    Azure::Core::IO::MemoryBodyStream pageStream(pageBytes.data(), pageBytes.size());
    pageBlob.UploadPages(0, pageStream);
    std::size_t rangeCount = 0;
    for (auto page = pageBlob.GetPageRanges(); page.HasPage(); page.MoveToNextPage())
    {
      for (const auto& range : page.PageRanges)
      {
        Require(range.Offset == 0, "page range start mismatch");
        Require(range.Length.HasValue() && range.Length.Value() == 512, "page range length mismatch");
        ++rangeCount;
      }
    }
    Require(rangeCount == 1, "expected exactly one page range");
    Require(DownloadRange(pageBlob, 0, 512) == pageBytes, "page blob download mismatch");

    Blobs::BlobLeaseClient lease(block, Blobs::BlobLeaseClient::CreateUniqueLeaseId());
    auto leaseResult = lease.Acquire(Blobs::BlobLeaseClient::InfiniteLeaseDuration).Value;
    Require(!leaseResult.LeaseId.empty(), "lease ID is missing");
    lease.Release();

    Azure::Storage::Sas::BlobSasBuilder sasBuilder;
    sasBuilder.Protocol = Azure::Storage::Sas::SasProtocol::HttpsAndHttp;
    sasBuilder.StartsOn = std::chrono::system_clock::now() - std::chrono::minutes(1);
    sasBuilder.ExpiresOn = std::chrono::system_clock::now() + std::chrono::minutes(10);
    sasBuilder.BlobContainerName = containerName;
    sasBuilder.BlobName = "folder/sdk-roundtrip.bin";
    sasBuilder.Resource = Azure::Storage::Sas::BlobSasResource::Blob;
    sasBuilder.SetPermissions(Azure::Storage::Sas::BlobSasPermissions::Read);
    const auto sasToken = sasBuilder.GenerateSasToken(
        Azure::Storage::StorageSharedKeyCredential(accountName, accountKey));
    Blobs::BlobClient sasBlob(block.GetUrl() + sasToken);
    Require(Download(sasBlob) == content, "SAS download mismatch");

    std::cout << "azure-storage-blobs-cpp " << SdkVersion
              << " compatibility passed for block, staged, append, page, snapshot, lease, list, "
                 "tags, ranges, and SAS operations.\n";
  }
  catch (...)
  {
    if (created)
    {
      try
      {
        container.Delete();
      }
      catch (...)
      {
      }
    }
    throw;
  }

  if (created)
  {
    container.Delete();
  }
  return 0;
}

File: Mk8.Sava.Tests/SdkCompatibility/go/main.go
package main

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"io"
	"os"
	"strings"
	"time"

	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
	"github.com/Azure/azure-sdk-for-go/sdk/azcore/streaming"
	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blob"
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blockblob"
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/container"
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/lease"
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/sas"
)

const sdkVersion = "v1.8.1"

func main() {
	ctx := context.Background()
	endpoint := requiredEnvironment("MK8_SAVA_BLOB_ENDPOINT")
	accountName := requiredEnvironment("MK8_SAVA_ACCOUNT_NAME")
	accountKey := requiredEnvironment("MK8_SAVA_ACCOUNT_KEY")
	credential := must(azblob.NewSharedKeyCredential(accountName, accountKey))
	clientOptions := &azblob.ClientOptions{ClientOptions: noRetryOptions()}
	client := must(azblob.NewClientWithSharedKeyCredential(endpoint, credential, clientOptions))
	containerName := "go-" + randomHex(16)
	created := false
	defer func() {
		if created {
			_, _ = client.DeleteContainer(ctx, containerName, nil)
		}
	}()

	must(client.CreateContainer(ctx, containerName, &azblob.CreateContainerOptions{
		Metadata: map[string]*string{"runtime": to.Ptr("go")},
	}))
	created = true

	containers := client.NewListContainersPager(&azblob.ListContainersOptions{
		Prefix:  to.Ptr(containerName),
		Include: azblob.ListContainersInclude{Metadata: true},
	})
	containerCount := 0
	for containers.More() {
		page := must(containers.NextPage(ctx))
		for _, item := range page.ContainerItems {
			require(dereference(item.Name) == containerName, "unexpected container name")
			require(findMetadata(item.Metadata, "runtime") == "go", "container metadata mismatch")
			containerCount++
		}
	}
	require(containerCount == 1, "expected exactly one container")

	content := make([]byte, 256*1024+43)
	for index := range content {
		content[index] = byte((index*37 + index/269) % 251)
	}
	containerClient := client.ServiceClient().NewContainerClient(containerName)
	blockClient := containerClient.NewBlockBlobClient("folder/sdk-roundtrip.bin")
	must(blockClient.UploadBuffer(ctx, content, &blockblob.UploadBufferOptions{
		HTTPHeaders: &blob.HTTPHeaders{
			BlobContentType:  to.Ptr("application/x-mk8-go"),
			BlobCacheControl: to.Ptr("private,max-age=23"),
		},
		Metadata: map[string]*string{"owner": to.Ptr("go-sdk")},
		Tags:     map[string]string{"runtime": "go", "purpose": "compatibility"},
	}))
	require(bytes.Equal(download(must(blockClient.DownloadStream(ctx, nil))), content), "block download mismatch")
	rangeResponse := must(blockClient.DownloadStream(ctx, &blob.DownloadStreamOptions{
		Range: blob.HTTPRange{Offset: 12_345, Count: 7_777},
	}))
	require(bytes.Equal(download(rangeResponse), content[12_345:20_122]), "range download mismatch")

	properties := must(blockClient.GetProperties(ctx, nil))
	require(dereference(properties.ContentLength) == int64(len(content)), "content length mismatch")
	require(dereference(properties.ContentType) == "application/x-mk8-go", "content type mismatch")
	require(dereference(properties.CacheControl) == "private,max-age=23", "cache control mismatch")
	require(findMetadata(properties.Metadata, "owner") == "go-sdk", "blob metadata mismatch")
	tags := must(blockClient.GetTags(ctx, nil))
	require(findTag(tags.BlobTagSet, "runtime") == "go", "blob tag mismatch")

	blobs := containerClient.NewListBlobsFlatPager(&container.ListBlobsFlatOptions{
		Prefix:  to.Ptr("folder/"),
		Include: container.ListBlobsInclude{Metadata: true, Tags: true},
	})
	blobCount := 0
	for blobs.More() {
		page := must(blobs.NextPage(ctx))
		for _, item := range page.Segment.BlobItems {
			require(dereference(item.Name) == "folder/sdk-roundtrip.bin", "listed blob name mismatch")
			require(findMetadata(item.Metadata, "owner") == "go-sdk", "listed metadata mismatch")
			require(findTag(item.BlobTags.BlobTagSet, "runtime") == "go", "listed tag mismatch")
			blobCount++
		}
	}
	require(blobCount == 1, "expected exactly one listed blob")

	snapshot := must(blockClient.CreateSnapshot(ctx, &blob.CreateSnapshotOptions{
		Metadata: map[string]*string{"owner": to.Ptr("snapshot")},
	}))
	snapshotClient := must(blockClient.WithSnapshot(dereference(snapshot.Snapshot)))
	require(bytes.Equal(download(must(snapshotClient.DownloadStream(ctx, nil))), content), "snapshot download mismatch")
	snapshotProperties := must(snapshotClient.GetProperties(ctx, nil))
	require(findMetadata(snapshotProperties.Metadata, "owner") == "snapshot", "snapshot metadata mismatch")

	stagedClient := containerClient.NewBlockBlobClient("staged.bin")
	blockIDs := []string{
		base64.StdEncoding.EncodeToString([]byte("block-0001")),
		base64.StdEncoding.EncodeToString([]byte("block-0002")),
	}
	first := []byte("first-block|")
	second := []byte("second-block")
	must(stagedClient.StageBlock(ctx, blockIDs[0], streaming.NopCloser(bytes.NewReader(first)), nil))
	must(stagedClient.StageBlock(ctx, blockIDs[1], streaming.NopCloser(bytes.NewReader(second)), nil))
	must(stagedClient.CommitBlockList(ctx, blockIDs, &blockblob.CommitBlockListOptions{
		Metadata: map[string]*string{"route": to.Ptr("staged")},
		Tags:     map[string]string{"runtime": "go"},
	}))
	require(
		bytes.Equal(download(must(stagedClient.DownloadStream(ctx, nil))), append(first, second...)),
		"staged block download mismatch",
	)
	blockList := must(stagedClient.GetBlockList(ctx, blockblob.BlockListTypeCommitted, nil))
	require(len(blockList.CommittedBlocks) == len(blockIDs), "committed block count mismatch")
	for index, item := range blockList.CommittedBlocks {
		require(dereference(item.Name) == blockIDs[index], "committed block ID mismatch")
	}
	require(len(blockList.UncommittedBlocks) == 0, "unexpected uncommitted blocks")

	appendClient := containerClient.NewAppendBlobClient("append.log")
	must(appendClient.Create(ctx, nil))
	must(appendClient.AppendBlock(ctx, streaming.NopCloser(bytes.NewReader([]byte("alpha"))), nil))
	must(appendClient.AppendBlock(ctx, streaming.NopCloser(bytes.NewReader([]byte("|beta"))), nil))
	require(
		bytes.Equal(download(must(appendClient.DownloadStream(ctx, nil))), []byte("alpha|beta")),
		"append blob download mismatch",
	)

	pageClient := containerClient.NewPageBlobClient("page.bin")
	must(pageClient.Create(ctx, 1024, nil))
	pageBytes := bytes.Repeat([]byte{0x5a}, 512)
	must(pageClient.UploadPages(
		ctx,
		streaming.NopCloser(bytes.NewReader(pageBytes)),
		blob.HTTPRange{Offset: 0, Count: int64(len(pageBytes))},
		nil,
	))
	pageRanges := pageClient.NewGetPageRangesPager(nil)
	rangeCount := 0
	for pageRanges.More() {
		page := must(pageRanges.NextPage(ctx))
		for _, allocated := range page.PageRange {
			require(dereference(allocated.Start) == 0, "page range start mismatch")
			require(dereference(allocated.End) == 511, "page range end mismatch")
			rangeCount++
		}
	}
	require(rangeCount == 1, "expected exactly one page range")
	pageDownload := must(pageClient.DownloadStream(ctx, &blob.DownloadStreamOptions{
		Range: blob.HTTPRange{Offset: 0, Count: 512},
	}))
	require(bytes.Equal(download(pageDownload), pageBytes), "page blob download mismatch")

	leaseClient := must(lease.NewBlobClient(blockClient, nil))
	leaseResponse := must(leaseClient.AcquireLease(ctx, -1, nil))
	require(leaseResponse.LeaseID != nil, "lease ID is missing")
	must(leaseClient.ReleaseLease(ctx, nil))

	start := time.Now().UTC().Add(-time.Minute)
	sasURL := must(blockClient.GetSASURL(
		sas.BlobPermissions{Read: true},
		time.Now().UTC().Add(10*time.Minute),
		&blob.GetSASURLOptions{StartTime: &start},
	))
	sasClient := must(blob.NewClientWithNoCredential(
		sasURL,
		&blob.ClientOptions{ClientOptions: noRetryOptions()},
	))
	require(bytes.Equal(download(must(sasClient.DownloadStream(ctx, nil))), content), "SAS download mismatch")

	fmt.Printf(
		"azblob %s compatibility passed for block, staged, append, page, snapshot, lease, list, tags, ranges, and SAS operations.\n",
		sdkVersion,
	)
}

func noRetryOptions() azcore.ClientOptions {
	return azcore.ClientOptions{Retry: policy.RetryOptions{MaxRetries: -1}}
}

func download(response blob.DownloadStreamResponse) []byte {
	defer response.Body.Close()
	return must(io.ReadAll(response.Body))
}

func findTag(tags []*blob.Tags, key string) string {
	for _, tag := range tags {
		if dereference(tag.Key) == key {
			return dereference(tag.Value)
		}
	}
	return ""
}

func findMetadata(metadata map[string]*string, key string) string {
	for name, value := range metadata {
		if strings.EqualFold(name, key) {
			return dereference(value)
		}
	}
	return ""
}

func randomHex(byteCount int) string {
	value := make([]byte, byteCount)
	must(rand.Read(value))
	return hex.EncodeToString(value)
}

func requiredEnvironment(name string) string {
	value := os.Getenv(name)
	if value == "" {
		panic("missing required environment variable " + name)
	}
	return value
}

func must[T any](value T, err error) T {
	if err != nil {
		panic(err)
	}
	return value
}

func dereference[T any](value *T) T {
	if value == nil {
		panic("unexpected nil SDK response field")
	}
	return *value
}

func require(condition bool, message string) {
	if !condition {
		panic(message)
	}
}
